apps/system/js/advanced_telemetry.js

```diff
@@ -324,6 +362,84 @@
     }
   };
 
+  // Check if there are existing metrics that need to be merged, if so,
+  // merge them and then transmit.
+  AT.prototype.mergeSend = function mergeSend(payloadNew, send) {
+    var self = this;
+    // Check if there are metrics in asyncStorage that need to be merged.
+    // If not, then we can just send them.
+    asyncStorage.getItem(AT.METRICS_KEY, function(value) {
+      if (value) {
+        mergeMetrics(value, payloadNew, self);
+      } else {
+        if (send) {
+          self.transmit(payloadNew);
+        } else {
+          asyncStorage.setItem(AT.METRICS_KEY, payloadNew);
+        }
+      }
+    });
+
+    // Merge the new metrics into the old metrics and then transmit the
+    // result to the server.
+    function mergeMetrics(payloadOld, payloadNew, self) {
+      mergeKeyed(payloadOld.payload.keyedHistograms, payloadOld, payloadNew);
+      mergeLists(payloadOld.payload.addonHistograms,
+                 payloadNew.payload.addonHistograms);
+      if (send) {
```

Comment:
More readable, cleaner to simply pass the old a new payload objects - the functions can pull out what they need from the payloads.